vllm/worker/model_runner.py

```diff
@@ -220,35 +220,115 @@ def __init__(
-            self.orig_seq_lens = orig_seq_lens or []
-            self.query_lens = query_lens or []
-            self.context_lens = context_lens or []
-            self.curr_sliding_window_blocks = curr_sliding_window_blocks or []
-
-            self.lora_index_mapping = lora_index_mapping or []
-            self.lora_prompt_mapping = lora_prompt_mapping or []
-            self.lora_requests = lora_requests or set()
-
-            self.prompt_adapter_index_mapping = (prompt_adapter_index_mapping
-                                                 or [])
-            self.prompt_adapter_prompt_mapping = (prompt_adapter_prompt_mapping
-                                                  or [])
-            self.prompt_adapter_request = prompt_adapter_request
 
+            if reinit:
+                if input_tokens:
+                    self.input_tokens = input_tokens
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.input_tokens[seq_id].clear()
+
+                if input_positions:
+                    self.input_positions = input_positions
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.input_positions[seq_id].clear()
+
+                if seq_lens:
+                    self.seq_lens = seq_lens
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.seq_lens[seq_id] = 0
+
+                if orig_seq_lens:
+                    self.orig_seq_lens = orig_seq_lens
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.orig_seq_lens[seq_id] = 0
+
+                if query_lens:
+                    self.query_lens = query_lens
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.query_lens[seq_id] = 0
+
+                if context_lens:
+                    self.context_lens = context_lens
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.context_lens[seq_id] = 0
+
+                if curr_sliding_window_blocks:
+                    self.curr_sliding_window_blocks = curr_sliding_window_blocks
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.curr_sliding_window_blocks[seq_id] = 0
+
+                if lora_index_mapping:
+                    self.lora_index_mapping = lora_index_mapping
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.lora_index_mapping[seq_id].clear()
+
+                if lora_prompt_mapping:
+                    self.lora_prompt_mapping = lora_prompt_mapping
+                else:
+                    for seq_id in range(len(self.seq_ids)):
+                        self.lora_prompt_mapping[seq_id].clear()
```

Comment:
I feel like the above could be simplified .. maybe collect all the things to be cleared and all the things to be zeroed and then have a single loop over the sequences? These lists themselves could possibly also be reused.

Reply:
This one is harder to generalize because it uses different fields. I tried to do lists, but then you still need to separate to different fields and it complicates things again. I will see if I can make the code cleaner here.

Reply:
Please note that we were planning to use dataclass with kw_only=True, but it is supported in Python 3.10 so we fallback to the current implementation. We should eventually use dataclass tho.